usr/opt/OSC200/usr/share/man/man1/lpq.1
.\" This manpage source uses rsml coding. 
.so /usr/share/lib/tmac/sml
.so /usr/share/lib/tmac/rsml
.\"
.\"
.\" *********************************************************************
.\" *                                                                   *
.\" *      Copyright (c) Digital Equipment Corporation, 1991, 1994      *
.\" *                                                                   *
.\" *                       All Rights Reserved.                        *
.\" *                                                                   *
.\" *********************************************************************
.\"
.\"
.\" HISTORY
.\"
.\" @(#)$RCSfile: lpq.1,v $ $Revision: 3.5.7.2 $ (DEC) $Date: 1993/09/28 19:53:37 $
.\"
.\" (c) Copyright 1990, 1991, 1992 OPEN SOFTWARE FOUNDATION, INC. 
.\" ALL RIGHTS RESERVED 
.\" OSF/1 Release 1.0.4
.\"build rmifdef
.TH lpq 1
.SH NAME
\*Llpq\*O \- Examines the spool queues.
.SH SYNOPSIS
.PP
.sS
\*Llpq\*O
[\*L+\*O][\*Vnumber\*O]
[\*L\-l\*O]
[\*L\-P\*Vprinter\*O]
[\*Vrequest-ID\*O
\&.\&.\&.]
[\*Vuser\*O
\&.\&.\&.]
.sE
.SH FLAGS
.PP
.VL 4m
.LI "\*L\-l\*O"
Displays information about each of the files comprising
the printer request.
.LI "\*L\-P\*Vprinter\*O"
Specifying \*Vprinter\*O displays requests queued for that printer only.
.LI "\*L\(pl\*O[\*Vnumber\*O]"
Displays the spool queue until it empties.
Sleeps \*Vnumber\*O seconds between scans of the queue if \*Vnumber\*O is
specified.
.LE
.SH DESCRIPTION
.PP
.iX "lpq command"
.iX "spooling" "printer queue examination"
.iX "queues" "examination"
.iX "printers" "queue examination"
The \*Llpq\*O command examines the spooling area used by
.gL "spooling"
\*Llpd\*O for printing files on the line printer and reports the status
of the specified requests or all requests associated with a user.
.PP
The \*Llpq\*O command invoked without any arguments
reports on any requests currently in the queue.
The \*L\-P\*O flag can be used to specify a particular printer; otherwise,
the default line printer is used (or the value of the \*LPRINTER\*O
environment variable).
.PP
Specifying one or more \*Vrequest-ID\*Os displays only the specified
requests.  Specifying one or more \*Vuser\*Os displays print requests from
those users only.
.PP
If you specify the \*L+\*O argument,
\*Llpq\*O displays the spool queue until it empties.  Supplying a
number immediately after the \*L+\*O argument indicates that \*Llpq\*O
should sleep \*Vnumber\*O seconds in between scans of the queue.  All
other arguments supplied are interpreted as \*Vuser\*O or \*Lrequest-ID\*O
arguments to display only those requests of interest.
.PP
For each request submitted (that is, each request invoked
by \*Llp\*O or \*Llpr\*O),
\*Llpq\*O reports the user's name, current rank in the queue, the
names of files comprising the request, the request ID (a number that
you can use to identify the request to other commands, such as \*Llprm\*O)
and the total size of the job in bytes.
.PP
The \*L\-l\*O flag causes information about each of the files comprising
the request to be displayed; without it, only as much information as will
fit on one line is displayed.  Job ordering depends on
the algorithm used to scan the spooling directory and is supposed
to be FIFO (First In First Out).  Filenames comprising a request might
.gL "FIFO"
.gL "First In First Out"
be unavailable (when \*Llp\*O or \*Llpr\*O is used as a sink in a pipeline),
in which case the file is indicated as follows:
.oS
standard input
.oE
.SH NOTES
.PP
Due to the dynamic nature of the information in the spooling
directory, \*Llpq\*O might report unreliably.  Output formatting is
sensitive to the line length of the terminal and can result in
widely spaced columns.
.SH EXAMPLES
.PP
To display a request in the print queue, enter:
.sp
.iS
lpq \-Pprinter1
.iE
.PP
This command displays a list similar to the following:
.oS
Mon March 2 15:15:15 1990: printer1 is ready and printing

Rank    Pri   Owner  Job    Files              Total Size
active   0    rcb    888    massachusetts      1024 bytes
1st      0    jmp    110    berkshire          2048 bytes
.oE
.PP
(Output may be formatted differently on your system.)
.PP
An appropriate message is displayed if the lock file is malformed.
.SH DIAGNOSTICS
.PP
Appropriate diagnostic messages appear under several conditions:
.ML
.LI
If there is no daemon active, but there are files in the spooling directory,
the message \*CWarning: no daemon present\*O is displayed.
.LI
If the network is unreachable, the message \*Cconnection to\*O \*Vdevice\*O
\*Cis down\*O is displayed.
.LI
If the printcap file is unavailable, the message \*Clpq:\*O
\*Vdevice\*O \*Ccannot open printer description file\*O is displayed.
.LI
If the spooling directory is unavailable, the message \*Clpq:\*O
\*Vdevice\*O \*Ccannot chdir to spooling directory\*O is displayed.
.LE
.SH FILES
.PP
.VL 10m
.LI "\*L/usr/share/lib/termcap\*O"
Manipulates the screen for repeated display.
.LI "\*L/etc/printcap\*O"
Printer description file.
.LI "\*L/var/spool/*\*O"
Spool directories.
.LI "\*L/var/spool/*/cf*\*O"
Daemon control files.
.LI "\*L/var/spool/*/df*\*O"
Data files specified in \*Lcf\*O files.
.LI "\*L/var/spool/*/tf*\*O"
Temporary copies of \*Lcf\*O files.
.LI "\*L/var/spool/*/lock\*O"
Lock file used to obtain the process ID of the current
daemon and the request ID of the currently active request.
.LE
.SH RELATED INFORMATION
.PP
Commands: \*Llp\*O(1)/\*Lcancel\*O(1), \*Llpc\*O(8), \*Llpd\*O(8),
\*Llpr\*O(1), \*Llprm\*O(1), \*Llpstat\*O(1)
.EQ
delim off
.EN


usr/opt/OSC200/src/usr/sbin/lpd/lpq.c
/*
 * *****************************************************************
 * *                                                               *
 * *    Copyright (c) Digital Equipment Corporation, 1991, 1994    *
 * *                                                               *
 * *   All Rights Reserved.  Unpublished rights  reserved  under   *
 * *   the copyright laws of the United States.                    *
 * *                                                               *
 * *   The software contained on this media  is  proprietary  to   *
 * *   and  embodies  the  confidential  technology  of  Digital   *
 * *   Equipment Corporation.  Possession, use,  duplication  or   *
 * *   dissemination of the software and media is authorized only  *
 * *   pursuant to a valid written license from Digital Equipment  *
 * *   Corporation.                                                *
 * *                                                               *
 * *   RESTRICTED RIGHTS LEGEND   Use, duplication, or disclosure  *
 * *   by the U.S. Government is subject to restrictions  as  set  *
 * *   forth in Subparagraph (c)(1)(ii)  of  DFARS  252.227-7013,  *
 * *   or  in  FAR 52.227-19, as applicable.                       *
 * *                                                               *
 * *****************************************************************
 */
/*
 * HISTORY
 */
#ifndef lint
static char	*sccsid = "@(#)$RCSfile: lpq.c,v $ $Revision: 4.2.5.2 $ (DEC) $Date: 1993/10/08 15:14:32 $";
#endif 
/*
 */
/*
 * (c) Copyright 1990, OPEN SOFTWARE FOUNDATION, INC.
 * ALL RIGHTS RESERVED
 */
/*
 * OSF/1 Release 1.0
 */
#if !defined(lint) && !defined(_NOIDENT)

#endif
/*
 * Copyright (c) 1983 Regents of the University of California.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms are permitted
 * provided that the above copyright notice and this paragraph are
 * duplicated in all such forms and that any documentation,
 * advertising materials, and other materials related to such
 * distribution and use acknowledge that the software was developed
 * by the University of California, Berkeley.  The name of the
 * University may not be used to endorse or promote products derived
 * from this software without specific prior written permission.
 * THIS SOFTWARE IS PROVIDED ``AS IS'' AND WITHOUT ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, WITHOUT LIMITATION, THE IMPLIED
 * WARRANTIES OF MERCHANTIBILITY AND FITNESS FOR A PARTICULAR PURPOSE.
 * 
 * lpq.c	5.5 (Berkeley) 6/30/88";
 * lpq.c	4.1 15:58:48 7/19/90 SecureWare
 */


/*
 * Spool Queue examination program
 *
 * lpq [-l] [-Pprinter] [user...] [job...]
 *
 * -l long output
 * -P used to identify printer as per lpr/lprm
 */

#include <locale.h>
#include "lp.h"

#if SEC_BASE
#define MSGSTR_SEC(n,s) catgets(catd,MS_PRINTER_SEC,n,s)
#endif SEC_BASE

#if SEC_BASE
#include <sys/security.h>
#endif

char	*user[MAXUSERS];	/* users to process */
int	users;			/* # of users in user array */
int	requ[MAXREQUESTS];	/* job number of spool entries */
int	requests;		/* # of spool requests */

main(argc, argv)
	register int	argc;
	register char	**argv;
{
    char *thisArg;
	int	ch, lflag;		/* long output option */
	int     seconds = 0;
        int displayq(int);

        (void) setlocale( LC_ALL, "" );
        catd = catopen(MF_PRINTER,NL_CAT_LOCALE);
#if SEC_BASE
	set_auth_parameters(argc, argv);
	initprivs();
#endif

	name = *argv++;
	if (gethostname(host, sizeof(host))) {
		perror(MSGSTR(LPQ_1, "lpq: gethostname"));
		exit(1);
	}
	openlog("lpd", 0, LOG_LPR);

	lflag = 0;
	while ( --argc > 0 )
	{
	    thisArg = *argv;
	    if ( *thisArg == '+' )
	    {
		if (thisArg[1])
		    seconds = atoi(thisArg);
		if ( seconds <= 0 )
		    seconds = 1;
		argv++;
	    }
	    else if ( *thisArg == '-' )
	    {
		switch(thisArg[1]) 
		{
		case 'l':			/* long output */
		    ++lflag;
		    break;
		case 'P':		/* printer name */
		    if ( thisArg[2] )
			printer = thisArg+2;
		    else
			if( --argc ) printer = *++argv;
		    break;
		case '?':
		default:
		    usage();
		    }
		argv++;
	    }
	    else
		break;
	}
	if (printer == NULL && (printer = getenv("PRINTER")) == NULL)
		printer = DEFLP;

	for (; argc > 0 ; --argc)
	{
	    thisArg = *argv++;
	    if (isdigit(*thisArg))
	    {
		if (requests >= MAXREQUESTS)
		    fatal(MSGSTR(LPQ_2, "too many requests"));
		requ[requests++] = atoi(thisArg);
	    }
	    else
	    {
		if (users >= MAXUSERS)
		    fatal(MSGSTR(LPQ_3, "too many users"));
		user[users++] = thisArg;
	    }
	}
#if SEC_BASE
    if (!forcepriv(SEC_REMOTE)
#if SEC_MAC
	|| authorized_user("macquery") && !forcepriv(SEC_ALLOWMACACCESS)
#endif
	)
    {
	fprintf(stderr, MSGSTR_SEC(PRIVS, "%s: insufficient privileges\n"), "lpq");
	exit(1);
    }
#endif
    if (seconds)
    {
	void clearscreen();
	clearscreen();
	while (displayq(lflag))
	{
	    sleep(seconds);
	    clearscreen();
	}
    }
    else
	displayq(lflag);
    exit(0);
}

void clearscreen()
{
#ifdef TERMCAP
    char    *tgetstr();
    char    PC;
    char *cp = getenv("TERM");
    char clbuf[20];
    char pcbuf[20];
    char *clbp = clbuf;
    char *pcbp = pcbuf;
    char *clear;
    char buf[1024];
    char *pc;
    
    if (cp == (char *) 0)
	return;
    if (tgetent(buf, cp) != 1)
	return;
    pc = tgetstr("pc", &pcbp);
    if (pc)
	PC = *pc;
    clear = tgetstr("cl", &clbp);
    if (clear)
	tputs(clear, tgetnum("li"), putchar);
    return;
#endif
}

usage()
{
	fprintf(stderr, 
		MSGSTR(LPQ_4, "usage: lpq [+n] [-l] [-Pprinter] [user ...] [job ...]\n"));
	exit(1);
}

